accepted/0034-context-config-parameter.md

```diff
@@ -0,0 +1,116 @@
+
+By default `context` will have no value, this is how we will avoid a breaking change.
+
+The `adduser` and `logout` commands currently are hardcoded to use `'user'` instead of referencing a variable, which is behavior I believe should remain as-is in order to avoid users accidentally saving credentials into project-level config files.
+
+### `npm exec`
+
+Implementation here is somewhat more involved. The logic found [here](https://github.com/npm/cli/blob/8806015fdd025f73ccf4001472370eafd3c5a856/lib/exec.js#L99-L117) would be expanded to create an instance of arborist, load your actual tree and iterate its inventory looking for a bin based on the rules detailed above.
+
+For project installed packages, if a bin is found, we would verify that the link and/or shim located in that location's bin directory points to the exact package we matched. If it does not, we would overwrite it with one that does, and then execute the freshly linked/shimmed bin.
+
+For the case of globally installed packages, there is no need to ensure the correct bin is linked as we will only link the bin of a directly installed global package. However, we would still apply the search logic here so that `npm exec --context=global -- typescript` will function as expected.
+
+Remote package behavior would remain exactly as it is today.
+
+When the `context` config is unset (the default), we will maintain the existing behavior of searching locally installed packages, then globally installed packages, then remote packages. The only change would be that the bin we run will be overwritten to the shallowest match in the user's actual tree before executing it, eliminating the current race condition that exists in the case of a conflicting bin in your tree.
+
+There is the potential that when run in interactive mode we can also display a list of everywhere we found something we _could_ run based on the user's request and prompt them to tell us which one.
+
+## Unresolved Questions and Bikeshedding
+
+~Is `where` the best name for this config? What if there are too many conflicts in dependencies?~ EDIT: Document updated to use `context` instead of `where`.
+
+What would the shortcut be? `-c` is currently the short form of `--call`.
+
+### `npm config`
+
+Should we prevent a user from setting protected fields like `_auth` or `_token` to a local config?
+
+Are there other fields that are undesirable at the local level?
+
+### `npm exec`
+
+Should we prompt when multiple matches are found? If we would prompt but are unable to, what should we do?
```

Comment:
Not sure what this means? I feel like this is already answered by the note, further up, outlining that we will follow the existing binary resolution pattern in `npm exec` (ref. https://github.com/npm/rfcs/pull/336/files#diff-e69970007bc5185a814912b41199ac9648ecfc70be343865a55d558f4c417028R56-R57)

Reply:
iirc there was some concern about which bin was linked being non-deterministic, but after a quick read through of bin-links it seems like that should only be possible if the user passed `--force` (or somehow we ended up reifying something deeper in the dep tree before something shallower, which shouldn't be possible since arborist sorts deps so shallowest are first) so maybe it's not something to worry about